azure_devops.md

```diff
@@ -0,0 +1,230 @@
+Note that each step is defined as either a "script" or a "task".
+
+**Scripts** specify a specific command to run, such as our `echo` command, or something more complex (including executing a script file!).
+
+**Tasks** refer to a predefined module, typically built-in & provided by Microsoft. These can be useful for common requirements, such as installing a particular package, or interacting with the Azure platform.
+
+[A list of Microsoft provided tasks can be viewed here.](https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/?view=azure-pipelines)
+
+Check you're working on the right branch and then commit & push this file.
+
+Moving to the Pipelines view in ADO, we can now register our pipeline; unlike GitHub / GitLab, the pipeline doesn't automatically run just because we created a file with a magic name.
+
+To create your pipeline:
+* Click "Create Pipeline"
+* Select "Azure Repos Git" as your code location.
+* Select your repository
+* It should then locate your pipeline file, and you can click "Run"
+
+After a few seconds, you should see it has paused and a warning appears! Grant the pipeline permission to access the Default agent pool that we assigned it.
+
+Now the pipeline will try to run and we see ... **a failure**!
+
+Click into the job, and you should see an issue like:
+> _No agent found in pool Default which satisfies the specified demands_
+
+Unlike GitHub/GitLab, ADO doesn't automatically offer a free tier for running jobs. They will generally grant access to a single free Microsoft-hosted pipeline if [you apply for it](https://learn.microsoft.com/en-us/azure/devops/pipelines/troubleshooting/troubleshoot-start?view=azure-devops#check-for-available-parallel-jobs), but this can take a few business days to approve so won't work for us here.
+
+> If you're curious why - a number of CI/CD tools have restricted their pipeline free tier over the last few years, seemingly due to a rise in abuse of those pipelines with [people using them to mine cryptocurrencies](https://devblogs.microsoft.com/devops/change-in-azure-pipelines-grant-for-public-projects/)!
+
+We will therefore need to host a runner ourselves to give our pipelines somewhere to execute. Thankfully, we can do that directly on a VM hosted in our Azure cloud too.
+
+### Step 4 - Set up a (self-hosted) Pipeline Runner
+
+> The guidance we'll provide is based on the instructions you can find by visiting:
+> * Project Settings (bottom left of the menu in ADO)
+> * Pipelines -> Agent Pools
+> * Click on the "Default" pool, and "New Agent"
+
+In order to create and register your self hosted agent, we'll need access to:
+* A PAT (Personal Access Token) for the agent to have permission to talk to ADO
+* Your ADO organisation URL
+
+You can generate the PAT by clicking on the User Settings (at the top right, by your initials) and visiting the Personal Access Tokens page. For simplicity here, we'll break best security practice and generate a PAT with "Full Access".
```

Comment:
There's technically another "User Settings" on the initials button that could confuse people:
![image](https://github.com/user-attachments/assets/9ce3fa59-31e1-48b2-a673-1b2206fcd3f6)

```suggestion
You can generate the PAT by clicking on the User Settings (at the top right, to the left of your initials) and visiting the Personal Access Tokens page. For simplicity here, we'll break best security practice and generate a PAT with "Full Access".
```